javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -116,14 +146,39 @@ private static int treeSize(int branchCount, int depthShift) {
         } else if (n <= 0) {
             return empty();
         } else {
-            final NodeModifier take = Arrays::copyTake;
-            final Object[] root = modifyLeaf(array, depthShift, (offset + n) - 1, take, take);
-            return new BitMappedTrie<>(root, offset, n, depthShift);
+            final int index = n - 1;
+            final Object[] root = arePointingToSameLeaf(index, length() - 1)
+                                  ? array
+                                  : modifyLeaf(array, depthShift, offset + index, Arrays::copyTake, Identity);
+            return new BitMappedTrie<T>(root, offset, n, depthShift).collapseHeight();
         }
     }
 
+    private boolean arePointingToSameLeaf(int i, int j) {
+        final boolean result = firstDigit(offset + i, BRANCHING_BASE) == firstDigit(offset + j, BRANCHING_BASE);
+        assert result == (getLeaf(i) == getLeaf(j));
+        return result;
+    }
+
+    /* drop root node while it has a single element */
+    private BitMappedTrie<T> collapseHeight() {
+        Object[] array = this.array;
+        int offset = this.offset;
+        int shift = depthShift;
+        for (; shift > 0; shift -= BRANCHING_BASE) {
+            final int skippedElements = array.length - 1;
+            if (skippedElements != digit(offset, shift)) {
+                break;
+            }
+            array = getAt(array, skippedElements);
+            offset -= treeSize(skippedElements, shift);
+        }
+        return (shift == depthShift) ? this
+                                     : new BitMappedTrie<>(array, offset, length(), shift);
+    }
+
     /* descend the tree from root to leaf, applying the given modifications along the way, returning the new root */
-    private static Object[] modifyLeaf(Object[] root, int depthShift, int index, NodeModifier node, NodeModifier leaf) {
+    private Object[] modifyLeaf(Object[] root, int depthShift, int index, NodeModifier node, NodeModifier leaf) {
```

Comment:
It could stay static. Is there a reason to make it an instance method? Then implicit `this` is also passed...